pkg/server/api/middleware/ratelimit.go

```diff
@@ -0,0 +1,221 @@
+// server-wide limit for a group of methods.
+func PerCallLimit(limit int) api.RateLimiter {
+	return newPerCallLimiter(limit)
+}
+
+// PerIPLimit returns a rate limiter that imposes a per-ip limit on calls
+// to a method. It can be shared across methods to enforce per-ip limits for
+// a group of methods.
+func PerIPLimit(limit int) api.RateLimiter {
+	return newPerIPLimiter(limit)
+}
+
+// WithRateLimits returns a middleware that performs rate limiting for the
+// group of methods descripted by the rateLimits map. It provides the
+// configured rate limiter to the method handlers via the request context. If
+// the middleware is invoked for a method is not described in the map, it will
+// fail the RPC with an INTERNAL error code, describing the RPC that was not
+// configured properly.  The middleware also encourages proper rate limiting by
+// logging errors if a handler fails to invoke the rate limiter provided on the
+// context when a limit has been configured or the handler invokes the rate
+// limiter when a no limit has been configured.
+//
+// The WithRateLimits middleware depends on the Logger and Authorization
+// middlewares.
+func WithRateLimits(rateLimits map[string]api.RateLimiter) Middleware {
+	return rateLimitsMiddleware{
+		limiters: rateLimits,
+	}
+}
+
+type noopLimiter struct{}
+
+func (noopLimiter) RateLimit(ctx context.Context, count int) error {
+	return nil
+}
+
+func (noopLimiter) Noop() bool { return true }
+
+func isNoopLimiter(rateLimiter api.RateLimiter) bool {
+	noop, ok := rateLimiter.(interface{ Noop() bool })
+	return ok && noop.Noop()
+}
+
+type perCallLimiter struct {
+	limiter rawRateLimiter
+}
+
+func newPerCallLimiter(limit int) *perCallLimiter {
+	return &perCallLimiter{limiter: newRawRateLimiter(rate.Limit(limit), limit)}
+}
+
+func (lim *perCallLimiter) RateLimit(ctx context.Context, count int) error {
+	return waitN(ctx, lim.limiter, count)
+}
+
+type perIPLimiter struct {
+	limit int
+
+	mtx      sync.RWMutex
+	limiters map[string]rawRateLimiter
+}
+
+func newPerIPLimiter(limit int) *perIPLimiter {
+	return &perIPLimiter{limit: limit,
+		limiters: make(map[string]rawRateLimiter),
+	}
+}
+
+func (lim *perIPLimiter) RateLimit(ctx context.Context, count int) error {
+	tcpAddr, ok := rpccontext.CallerAddr(ctx).(*net.TCPAddr)
+	if !ok {
+		// Calls not via TCP/IP aren't limited
+		return nil
+	}
+	limiter := lim.getLimiter(tcpAddr.IP.String())
+	return waitN(ctx, limiter, count)
+}
+
+func (lim *perIPLimiter) getLimiter(addr string) rawRateLimiter {
+	lim.mtx.RLock()
+	limiter, ok := lim.limiters[addr]
+	if ok {
+		lim.mtx.RUnlock()
+		return limiter
+	}
+	lim.mtx.RUnlock()
+
+	lim.mtx.Lock()
+	defer lim.mtx.Unlock()
+	limiter, ok = lim.limiters[addr]
+	if !ok {
+		limiter = newRawRateLimiter(rate.Limit(lim.limit), lim.limit)
+		lim.limiters[addr] = limiter
+	}
+
+	return limiter
+}
+
+type rateLimitsMiddleware struct {
+	limiters map[string]api.RateLimiter
```

Comment:
can this be accessed concurrently? should we have a mutex?

Reply:
It shouldn't be mutated after passing it into rate limits, so concurrent access should be safe. I'll document the intended ownership on the `WithRateLimits` call.